src/orion/core/io/config.py

```diff
@@ -0,0 +1,278 @@
+            If the configuration has subconfiguration, the key may be
+            hierarchical with each levels seperated by dots.
+            Ex: 'first.second.third'
+        value: object
+            A general object to set an option.
+
+        """
+        keys = key.split(".")
+        # Recursively in sub configurations
+        if len(keys) > 1:
+            subconfig = getattr(self, keys[0])
+            if subconfig is None:
+                raise KeyError("'{}' is not defined in configuration.".format(keys[0]))
+            subconfig[".".join(keys[1:])] = value
+        # Set in current configuration
+        else:
+            setattr(self, keys[0], value)
+
+    def __getitem__(self, key):
+        """Get option value using dict-like syntax
+
+        Parameters
+        ----------
+        key: str
+            The key or namespace to set the value of the configuration.
+            If the configuration has subconfiguration, the key may be
+            hierarchical with each levels seperated by dots.
+            Ex: 'first.second.third'
+
+        """
+        keys = key.split(".")
+        # Recursively in sub configurations
+        if len(keys) > 1:
+            subconfig = getattr(self, keys[0])
+            if subconfig is None:
+                raise KeyError("'{}' is not defined in configuration.".format(keys[0]))
+            return subconfig[".".join(keys[1:])]
+        # Set in current configuration
+        else:
+            return getattr(self, keys[0])
+
+    def add_option(self, key, option_type, default=NOT_SET, env_var=None):
+        """Add a configuration setting.
+
+        Parameters
+        ----------
+        key : str
+            The name of the configuration setting. This must be a valid
+            Python attribute name i.e. alphanumeric with underscores.
+        option_type : function
+            A function such as ``float``, ``int`` or ``str`` which takes
+            the configuration value and returns an object of the correct
+            type.  Note that the values retrieved from environment
+            variables are always strings, while those retrieved from the
+            YAML file might already be parsed. Hence, the function provided
+            here must accept both types of input.
+        default : object, optional
+            The default configuration to return if not set. By default none
+            is set and an error is raised instead.
+        env_var : str, optional
+            The environment variable name that holds this configuration
+            value. If not given, this configuration can only be set in the
+            YAML configuration file.
+
+        """
+        self.config[key] = {'type': option_type}
+        if env_var is not None:
+            self.config[key]['env_var'] = env_var
+        if default is not NOT_SET:
+            self.config[key]['default'] = default
+
+    # def to_dict(self):
```

Comment:
Should we remove the rest of this commented code?

Reply:
Yes!